File: Stripe/STPPaymentMethodAddress.m
//
//  STPPaymentMethodAddress.m
//  Stripe
//
//  Created by Yuki Tokuhiro on 3/5/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

#import "STPPaymentMethodAddress.h"

#import "NSDictionary+Stripe.h"
#import "STPAddress.h"

@interface STPPaymentMethodAddress ()

@property (nonatomic, copy, nonnull, readwrite) NSDictionary *allResponseFields;

@end

@implementation STPPaymentMethodAddress

- (instancetype)initWithAddress:(STPAddress *)address {
    self = [super init];
    if (self) {
        _city = [address.city copy];
        _country = [address.country copy];
        _line1 = [address.line1 copy];
        _line2 = [address.line2 copy];
        _postalCode = [address.postalCode copy];
        _state = [address.state copy];
    }
    return self;
}

- (NSString *)description {
    NSArray *props = @[
                       // Object
                       [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],
                       
                       // Properties
                       [NSString stringWithFormat:@"line1 = %@", self.line1],
                       [NSString stringWithFormat:@"line2 = %@", self.line2],
                       [NSString stringWithFormat:@"city = %@", self.city],
                       [NSString stringWithFormat:@"state = %@", self.state],
                       [NSString stringWithFormat:@"postalCode = %@", self.postalCode],
                       [NSString stringWithFormat:@"country = %@", self.country],
                       ];
    
    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

#pragma mark - STPFormEncodable

@synthesize additionalAPIParameters;

+ (nonnull NSDictionary *)propertyNamesToFormFieldNamesMapping {
    return @{
             NSStringFromSelector(@selector(line1)): @"line1",
             NSStringFromSelector(@selector(line2)): @"line2",
             NSStringFromSelector(@selector(city)): @"city",
             NSStringFromSelector(@selector(country)): @"country",
             NSStringFromSelector(@selector(state)): @"state",
             NSStringFromSelector(@selector(postalCode)): @"postal_code",
             };
}

+ (nullable NSString *)rootObjectName {
    return nil;
}

#pragma mark - STPAPIResponseDecodable

+ (nullable instancetype)decodedObjectFromAPIResponse:(nullable NSDictionary *)response {
    NSDictionary *dict = [response stp_dictionaryByRemovingNulls];
    if (!dict) {
        return nil;
    }
    STPPaymentMethodAddress *address = [self new];
    address.allResponseFields = dict;
    address.city = [dict stp_stringForKey:@"city"];
    address.country = [dict stp_stringForKey:@"country"];
    address.line1 = [dict stp_stringForKey:@"line1"];
    address.line2 = [dict stp_stringForKey:@"line2"];
    address.postalCode = [dict stp_stringForKey:@"postal_code"];
    address.state = [dict stp_stringForKey:@"state"];
    return address;
}

@end

File: Stripe/STPPaymentMethodAddress.swift
//
//  STPPaymentMethodAddress.swift
//  Stripe
//
//  Created by Yuki Tokuhiro on 3/5/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

import Contacts
import Foundation

/// The billing address, a property on `STPPaymentMethodBillingDetails`
public class STPPaymentMethodAddress: NSObject, STPAPIResponseDecodable, STPFormEncodable {
  @objc public var additionalAPIParameters: [AnyHashable: Any] = [:]

  /// City/District/Suburb/Town/Village.
  @objc public var city: String?
  /// 2-letter country code.
  @objc public var country: String?
  /// Address line 1 (Street address/PO Box/Company name).
  @objc public var line1: String?
  /// Address line 2 (Apartment/Suite/Unit/Building).
  @objc public var line2: String?
  /// ZIP or postal code.
  @objc public var postalCode: String?
  /// State/County/Province/Region.
  @objc public var state: String?

  /// Convenience initializer for creating a STPPaymentMethodAddress from an STPAddress.
  @objc
  public init(address: STPAddress) {
    super.init()
    city = address.city
    country = address.country
    line1 = address.line1
    line2 = address.line2
    postalCode = address.postalCode
    state = address.state
  }

  /// :nodoc:
  @objc public required override init() {
    super.init()
  }

  /// :nodoc:
  private(set) public var allResponseFields: [AnyHashable: Any] = [:]

  /// :nodoc:
  @objc public override var description: String {
    let props = [
      // Object
      String(format: "%@: %p", NSStringFromClass(STPPaymentMethodAddress.self), self),
      // Properties
      "line1 = \(line1 ?? "")",
      "line2 = \(line2 ?? "")",
      "city = \(city ?? "")",
      "state = \(state ?? "")",
      "postalCode = \(postalCode ?? "")",
      "country = \(country ?? "")",
    ]

    return "<\(props.joined(separator: "; "))>"
  }

  // MARK: - STPFormEncodable

  @objc
  public class func propertyNamesToFormFieldNamesMapping() -> [String: String] {
    return [
      NSStringFromSelector(#selector(getter:line1)): "line1",
      NSStringFromSelector(#selector(getter:line2)): "line2",
      NSStringFromSelector(#selector(getter:city)): "city",
      NSStringFromSelector(#selector(getter:country)): "country",
      NSStringFromSelector(#selector(getter:state)): "state",
      NSStringFromSelector(#selector(getter:CNMutablePostalAddress.postalCode)): "postal_code",
    ]
  }

  @objc
  public class func rootObjectName() -> String? {
    return nil
  }

  // MARK: - STPAPIResponseDecodable
  @objc
  public class func decodedObject(fromAPIResponse response: [AnyHashable: Any]?) -> Self? {
    guard let response = response else {
      return nil
    }
    let dict = (response as NSDictionary).stp_dictionaryByRemovingNulls() as NSDictionary
    let address = self.init()
    address.allResponseFields = response
    address.city = dict.stp_string(forKey: "city")
    address.country = dict.stp_string(forKey: "country")
    address.line1 = dict.stp_string(forKey: "line1")
    address.line2 = dict.stp_string(forKey: "line2")
    address.postalCode = dict.stp_string(forKey: "postal_code")
    address.state = dict.stp_string(forKey: "state")
    return address
  }
}
